internal/mesh/internal/controllers/sidecarproxy/builder/local_app.go

```diff
@@ -135,60 +135,85 @@ func destinationRulesByPort(allPorts []string, destinationRules []*pbauth.Destin
 		for _, p := range allPorts {
 			out[p] = nil
 		}
-
 		return out
 	}
 
 	for _, destinationRule := range destinationRules {
-		ports, dr := convertDestinationRule(allPorts, destinationRule)
-		for _, p := range ports {
-			if dr == nil {
+		portRules := convertDestinationRule(allPorts, destinationRule)
+		for p, pr := range portRules {
+			if pr.rule == nil {
 				out[p] = nil
 				continue
 			}
-			out[p] = append(out[p], dr)
+			out[p] = append(out[p], pr.rule)
 		}
 	}
 
 	return out
 }
 
-//nolint:unparam
-func convertDestinationRule(allPorts []string, dr *pbauth.DestinationRule) ([]string, *pbproxystate.DestinationRule) {
-	ports := make(map[string]struct{})
-	if len(dr.PortNames) > 0 {
-		for _, p := range dr.PortNames {
-			ports[p] = struct{}{}
-		}
-	} else {
-		for _, p := range allPorts {
-			ports[p] = struct{}{}
-		}
-	}
+type PortRule struct {
+	rule *pbproxystate.DestinationRule
+}
 
-	for _, exclude := range dr.Exclude {
-		for _, p := range exclude.PortNames {
-			delete(ports, p)
+func convertDestinationRule(allPorts []string, dr *pbauth.DestinationRule) map[string]*PortRule { // was []string, dr
+	portRules := make(map[string]*PortRule)
+	targetPorts := allPorts
+	if len(dr.PortNames) > 0 {
+		targetPorts = dr.PortNames
+	}
+	for _, p := range targetPorts {
+		if dr.PortsOnly() {
+			portRules[p] = &PortRule{}
+			for _, exclude := range dr.Exclude {
+				for _, ep := range exclude.PortNames {
+					delete(portRules, ep)
+				}
+			}
+		} else {
+			portRules[p] = makePortRule(dr, p)
 		}
 	}
+	return portRules
+}
 
-	var out []string
-	for p := range ports {
-		out = append(out, p)
-	}
-
-	if len(dr.String()) == 0 {
-		return out, nil
-	}
-
+func makePortRule(dr *pbauth.DestinationRule, p string) *PortRule {
 	psdr := &pbproxystate.DestinationRule{
 		PathExact:  dr.PathExact,
 		PathPrefix: dr.PathPrefix,
 		PathRegex:  dr.PathRegex,
 		Methods:    dr.Methods,
 	}
-	hrs := make([]*pbproxystate.DestinationRuleHeader, len(dr.Headers))
-	for i, hr := range dr.Headers {
+	psdr.DestinationRuleHeader = destinationRuleHeaders(dr.Headers)
+
+	var excls []*pbproxystate.ExcludePermissionRule
+	for _, ex := range dr.Exclude {
+		if len(ex.PortNames) == 0 || listContains(ex.PortNames, p) {
+			excls = append(excls, &pbproxystate.ExcludePermissionRule{
+				PathExact:  ex.PathExact,
+				PathPrefix: ex.PathPrefix,
+				PathRegex:  ex.PathRegex,
+				Methods:    ex.Methods,
+				Headers:    destinationRuleHeaders(ex.Headers),
+			})
+		}
+	}
+	psdr.Exclude = excls
+	return &PortRule{psdr}
+}
+
+func listContains(list []string, str string) bool {
+	for _, item := range list {
+		if item == str {
+			return true
+		}
+	}
+	return false
+}
+
+func destinationRuleHeaders(headers []*pbauth.DestinationRuleHeader) []*pbproxystate.DestinationRuleHeader {
+	hrs := make([]*pbproxystate.DestinationRuleHeader, len(headers))
```

Comment:
```suggestion
	hrs := make([]*pbproxystate.DestinationRuleHeader, 0, len(headers))
```

Reply:
I think that if you don't set the capacity the slice will have len == capacity. In this case that should be fine since we aren't optionally adding anything to the loops and we should have a full slice at the end.

In other cases though, if we were to only add headers on some condition and could have len(headers) < length of the allocated slice the rest of the range would be the empty value. I think that due to this its generally a good idea to set the capacity to 0 when allocation slices.

Reply:
if we were appending instead of directly using the index to set all values, I agree this would be the right approach